Draft for a .NET implementation of named tracers.

z1c0 opened the pull request:
Add interface `ITracerFactory` and default implementation `TracerFactory`.

The sample application `LoggingTracer.Demo.AspNetCore` makes use of `TracerIFactory` by providing an implementation (`LoggingTracerFactory` ) to the ASP.NET Core DI mechanism.

ZieglerWolfgang committed:
Draft for a .NET implementation of named tracers.

discostu105 commented on src/OpenTelemetry.Collector.AspNetCore/RequestsCollector.cs:
```diff
@@ -34,16 +34,16 @@ public class RequestsCollector : IDisposable
                     { "Microsoft.AspNetCore", (t, s) => new HttpInListener(t, s) },
                 },
-                tracer,
+                tracerFactory.Create("Microsoft.AspNetCore"),
```
shouldn't the name be "OpenTelemetry.Collector.AspNetCore" a.k.a. name of the instrumentation lib?

discostu105 commented on samples/LoggingTracer/LoggingTracer/LoggingTracerFactory.cs:
```diff
@@ -0,0 +1,20 @@
+            // - or return Tracer from a map for the the given name.
+            // - or return (new) Tracer based on a condition (e..g config) otherwise NopTracer, ... 
+            // - ...
+            return new LoggingTracer();
```
We could pass the tracer name to the LoggingTracer instance. It could use the name as logging prefix.

discostu105 commented on src/OpenTelemetry/Trace/Tracing.cs:
```diff
@@ -44,12 +43,18 @@ internal Tracing()
         /// Gets the tracer to record spans.
         /// </summary>
-        public static ITracer Tracer => (ITracer)tracer;
+        public static ITracer Tracer => tracer; // TODO: We could probably get rid of this property in favor of TracerFactory.
```
Yes, I'd say we should get rid of this. However, we might still need it for compat reasons :/.

Concerning "startEndHandler, TraceConfig" properties that the global Tracer currently receives. Aren't these properties that the TracerFactory should also get?

z1c0 replied:
Agreed. Adressed in follow-up commit.

discostu105 commented on test/OpenTelemetry.Collector.Dependencies.Tests/HttpClientTests.cs:
```diff
@@ -100,11 +100,13 @@ public async Task HttpOutCallsAreCollectedSuccessfullyAsync(HttpOutTestCase tc)
 
             var startEndHandler = new Mock<IStartEndHandler>();
             var tracer = new Tracer(startEndHandler.Object, new TraceConfig());
```
is the `tracer` instance still needed?

z1c0 replied:
It is returned be the mocked `TracerFactory`

ZieglerWolfgang committed:
Implement review feedback.
- LoggingTracer uses tracer name in log messages
- Remove global Tracer instance in favor of TracerFactory (Tracing.cs)
and update usages.
- Change tracer names to actually reflect the tracing library.

discostu105 commented on src/OpenTelemetry/Trace/Tracing.cs:
```diff
@@ -44,12 +43,18 @@ internal Tracing()
+        /// </summary>
+        public static ITracerFactory TracerFactory => tracerFactory;
+        
+/*        /// <summary>   
```
remove commented code?
